[WIP] Enable the transform tensor API to convert any NCHW tensor to NHWC tensor and vice versa

dmikushin opened the pull request:
LWPMIOPEN-125

We substitute the TensorTransfom implementation with a more generic GPU-based gpuTT library calls.

Few things still to be implemented:

* Support non-equal input and output types
* Support for alpha and beta in the case of a simple copy (i.e., perm = identity) is still missing
* The test should randomly transpose the input tensors, so that TensorTransform could transpose them back

dmikushin committed:
WIP Replacing TensorTransform actual code with a generic gpuTT-based implementation

dmikushin committed:
Not all layouts start with 'N', removing this check to support e.g. CHWN

dmikushin committed:
Using the most recent CMake in the Dockerfile, in order to have HIP and ROCm support in CMake

dmikushin committed:
Formatting source with clang-format-12, as suggested by @junliume

dmikushin committed:
Correcting sorting comparator

atamazov commented on src/ocl/tensorocl.cpp:
```diff
@@ -2120,6 +2122,28 @@ void CastTensor(const Handle& handle,
+        return a.first < b.first;
+    });
+
+    printf("Permutation for %s -> %s: ", src_layout.c_str(), dst_layout.c_str());
```
Use `MIOPEN_LOG_I2/I/W/E` macros, no printf please.

atamazov reviewed (changes requested):
I guess that gputt cannot be used with OCL backend. If so then please use conditional compilation (see config.h) as needed.

dmikushin committed:
No need to define __HIPCC__

dmikushin committed:
Leverage gputt only if the backend is HIP

dmikushin committed:
Find gputt package

dmikushin committed:
Checking in an addenum to the miopen:ci container that is needed to install the extra gputt dependency. It should be built like this:

  BUILDKIT_PROGRESS=plain docker build --no-cache -f Dockerfile.gputt -t rocm/miopen:ci-gputt .

Then the usual miopen:ci invocation should instead use miopen:ci-gputt:

  docker run -it -v $HOME:$HOME --privileged --rm --device=/dev/kfd --device /dev/dri:/dev/dri:rw  --volume /dev/dri:/dev/dri:rw -v /var/lib/docker/:/var/lib/docker --group-add video --cap-add=SYS_PTRACE --security-opt seccomp=unconfined rocm/miopen:ci-gputt fish

dmikushin committed:
Do not use hipcc, it's obsolete by direct use of /opt/rocm/llvm/bin/clang++

dmikushin committed:
Clone specifically a gputt branch tagged v0.1

dmikushin committed:
Update driver/transform_driver.hpp

Co-authored-by: Vasilii Filippov <DrizztDoUrden@users.noreply.github.com>

dmikushin committed:
Adding CMake and some tools that ease my development

dmikushin committed:
Merging requirements.txt

dmikushin committed:
Fixing warnings spotted by CI

dmikushin committed:
Adding driver option to test the tensor transform

dmikushin committed:
Acknowledging Xoffset and Yoffset that have to be applied to the vector pointers

dmikushin committed:
Getting to the point when we can completely replace all cases of tensor transform by gpuTT functionality. In order for that gpuTT has been extended to support all data types required by MIOpen, including the fancy int8x4

dmikushin committed:
Merge remote-tracking branch 'origin/develop' into enable-hchw-nhwc-conversion

dmikushin committed:
Bumping gputt to version with int8x4 support

dmikushin committed:
Merge remote-tracking branch 'origin/develop' into enable-hchw-nhwc-conversion

dmikushin committed:
Adding missing breaks, adding miopenInt8x4

dmikushin committed:
Forwarding gpuTT errors to MIOpen, forwarding MIOpen stream to gpuTT

dmikushin committed:
Merge remote-tracking branch 'origin/develop' into enable-hchw-nhwc-conversion

BradPepersAMD commented:
Closing this since we're moving MIOpen to the new monorepo approach you can see [here](https://github.com/ROCm/rocm-libraries).  Please re-submit this if it's still relevant.
